README.md
[![ CI](https://github.com/RailSafiullin/BallMoveSimulation/actions/workflows/app.yml/badge.svg)](https://github.com/RailSafiullin/BallMoveSimulation/actions/workflows/app.yml)
[![License](https://img.shields.io/badge/License-Apache_2.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
# Лабораторная работа по дисциплине "Компьютерная графика" №2

## Content

- [Preview](#preview)

- [How to install](#how-to-install-and-start)

- [Task 3](#task-3)

-   [License](#license)


## Preview

Application for simulating the motion of an elastic ball with specified physical parameters

Image of the application's initial window

![Watch the video](./img/initWindow.png)

Ball movement animation

![Watch the video](./img/animationApp.gif)

## How to install and start

```cmd

git clone https://github.com/RailSafiullin/BallMoveSimulation
cd BallMoveSimulation
pip install pytest
pip install pygame
python main.py

```

## Task 3

Требуется составить программу, которая демонстрирует процесс падения упругого шарика с заданной высоты на абсолютно твердую горизонтальную плоскость с последующими отражениями. Коэффициент отражения равен k. Программа должна завершаться по нажатию клавиши ESC или когда высота поднятия отскакивающего шарика станет меньше заданной величины h

## License

Distributed under the Apache License 2.0. See LICENSE.txt for more information.





CHANGELOG.md
# Changelog

## [1.0.0] - 2023-12-02

### Added
- Storage.py
- Controller.py
- Factory.py
- docs: help.txt, version.txt

### Update
- InitWindow
- main.py


## [0.1.0] - 2023-12-01

### Added

- Ball: add check collision function
- BallMoveSimulation: added exit from the program if the ball rises below the specified height


## [0.0.1] - 2023-12-01

### Added

- CHANGELOG.md
- main.py
- ball.py
- ballMoveSimulation.py

src/tests/test.py
import pytest
from Factory import Factory
from initWindow import InitWindow
from Storage import Storage
from ballMoveSimulation import BallMoveSimulation


@pytest.fixture(scope="module")
def app():
    app = Factory.create(InitWindow)
    yield app


@pytest.fixture(scope="module")
def storage():
    storage = Storage()
    yield storage


def test_storage_get_version(storage):
    result = storage.get_version()
    type(result) == str


def test_storage_get_help(storage):
    result = storage.get_help()
    type(result) == str


@pytest.fixture
def simulation():
    return BallMoveSimulation()

def test_init(simulation):
    assert simulation.WINDOW_WIDTH == 900
    assert simulation.WINDOW_HEIGHT == 600
    #assert simulation.ball_x == 450
    #assert simulation.ball_y == 20
    assert simulation.BALL_RADIUS == 20
    assert simulation.BALL_COLOR == (25, 25, 105)
    assert simulation.ball_vx == 0.0
    assert simulation.ball_vy == 0.0
    assert simulation.ball_ax == 0.0 
    assert simulation.ball_ay == 9.8 / simulation.clock_tick 
    assert simulation.k == 1.0
    assert simulation.clock_tick == 120
    assert simulation.escape_height == 10.0

def test_true_check_exit_condition(simulation):
    simulation.ball.y = 570
    simulation.ball.vy = -0.05
    simulation.ball.radius = 20
    simulation.escape_height = 10.0
    assert simulation.check_exit_condition() == True

def test_false_check_exit_condition(simulation):
    simulation.ball.y = 570
    simulation.ball.vy = -2.0
    simulation.ball.radius = 20
    simulation.escape_height = 10.0
    assert simulation.check_exit_condition() == False

def test_false2_check_exit_condition(simulation):
    simulation.ball.y = 500
    simulation.ball.vy = -0.05
    simulation.ball.radius = 20
    simulation.escape_height = 10.0
    assert simulation.check_exit_condition() == False

scr/tests/test_sample.py
import pytest
from Factory import Factory
from InitWindow import InitWindow
from Storage import Storage
from BallMoveSimulation import BallMoveSimulation


@pytest.fixture(scope="module")
def app():
    app = Factory.create(InitWindow)
    yield app


@pytest.fixture(scope="module")
def storage():
    storage = Storage()
    yield storage


def test_storage_get_version(storage):
    result = storage.get_version()
    type(result) == str


def test_storage_get_help(storage):
    result = storage.get_help()
    type(result) == str


@pytest.fixture
def simulation():
    return BallMoveSimulation()

def test_init(simulation):
    assert simulation.WINDOW_WIDTH == 900
    assert simulation.WINDOW_HEIGHT == 600
    assert simulation.ball_x == 450
    assert simulation.ball_y == 20
    assert simulation.BALL_RADIUS == 20
    assert simulation.BALL_COLOR == (25, 25, 105)
    assert simulation.ball_vx == 0.0
    assert simulation.ball_vy == 0.0
    assert simulation.ball_ax == 0.0
    assert simulation.ball_ay == 9.8
    assert simulation.k == 1.0
    assert simulation.clock_tick == 120
    assert simulation.escape_height == 10.0

def test_true_check_exit_condition(simulation):
    simulation.ball.y = 570
    simulation.ball.y = 0.05
    simulation.ball.radius = 20
    simulation.escape_height = 10.0
    assert simulation.check_exit_condition() == True

def test_false_check_exit_condition(simulation):
    simulation.ball.y = 570
    simulation.ball.y = 2.0
    simulation.ball.radius = 20
    simulation.escape_height = 10.0
    assert simulation.check_exit_condition() == False

scr/ball.py
import pygame


# Define the Ball class
class Ball:
    def __init__(self, x, y, radius, color, ball_vx, ball_vy, ball_ax, ball_ay, screen, k):
        self.x = x
        self.y = y
        self.radius = radius
        self.color = color
        self.screen = screen
        self.vx = ball_vx
        self.vy = ball_vy
        self.ax = ball_ax
        self.ay = ball_ay
        self.k = k

    def draw(self):
        pygame.draw.circle(self.screen, self.color, (int(self.x), int(self.y)), self.radius)

    def update(self):
        self.x += self.vx
        self.y += self.vy
        self.vx += self.ax
        if self.y + self.radius < self.screen.get_height():
            self.vy += self.ay
        self.check_collision()

    def check_collision(self):
        if self.y + self.radius > self.screen.get_height():
            self.y = self.screen.get_height() - self.radius
            self.vy = -self.k * self.vy
            if abs(self.vy) < 0.1:
                self.vy = 0
        if self.y - self.radius < 0:
            self.y = self.radius
            self.vy = -self.k * self.vy
        if self.x + self.radius > self.screen.get_width():
            self.x = self.screen.get_width() - self.radius
            self.vx = -self.k * self.vx
            if abs(self.vx) < 0.1:
                self.vx = 0
        if self.x - self.radius < 0:
            self.x = self.radius
            self.vx = -self.k * self.vx
            if abs(self.vx) < 0.1:
                self.vx = 0

src/ballMoveSimulation.py
import pygame
from Ball import Ball
import sys
class BallMoveSimulation:
    def __init__(self, WINDOW_WIDTH=900, WINDOW_HEIGHT=600,
                ball_x=450, ball_y=20, BALL_RADIUS=20,
                BALL_COLOR=(25, 25, 105),
                ball_vx=0.0, ball_vy=0.0, ball_ax=0.0, ball_ay=9.8, k=1.0,
                clock_tick=120, escape_height=10.0):
        self.WINDOW_WIDTH = WINDOW_WIDTH
        self.WINDOW_HEIGHT = WINDOW_HEIGHT
        self.clock_tick = clock_tick
        self.ball_x = ball_x
        self.ball_y = ball_y
        self.BALL_RADIUS = BALL_RADIUS
        self.BALL_COLOR = BALL_COLOR
        self.ball_vx = ball_vx
        self.ball_vy = ball_vy
        self.ball_ax = ball_ax / self.clock_tick
        self.ball_ay = ball_ay / self.clock_tick
        self.k = k
        self.escape_height = escape_height

        self.ball_x = self.WINDOW_WIDTH // 2
        self.ball_y = self.WINDOW_HEIGHT // 4

        pygame.init()

        self.simulation_window = pygame.display.set_mode((self.WINDOW_WIDTH, self.WINDOW_HEIGHT))
        pygame.display.set_caption("Elastic Ball Simulation")
        self.clock = pygame.time.Clock()

        self.ball = Ball(self.ball_x, self.ball_y, self.BALL_RADIUS, self.BALL_COLOR, self.ball_vx, self.ball_vy,
                        self.ball_ax, self.ball_ay, self.simulation_window, self.k)
    def start_simulation(self):
        while True:
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    pygame.quit()
                    sys.exit()

            keys = pygame.key.get_pressed()
            if keys[pygame.K_ESCAPE]:
                pygame.quit()
                sys.exit()
            if keys[pygame.K_UP]:
                self.ball.vy -= 0.1 + self.ball.ay
            if keys[pygame.K_DOWN]:
                self.ball.vy += 0.1
            if keys[pygame.K_LEFT]:
                self.ball.vx -= 0.1
            if keys[pygame.K_RIGHT]:
                self.ball.vx += 0.1

            self.simulation_window.fill((225, 225, 225))
            self.ball.update()

            if self.check_exit_condition():
                pygame.quit()
                sys.exit()

            self.ball.draw()
            pygame.display.update()
            self.clock.tick(self.clock_tick)

    def check_exit_condition(self):
        if self.ball.y + self.ball.radius >= self.WINDOW_HEIGHT - self.escape_height and (abs(self.ball.vy) < 0.1 and self.ball.vy <= 0):
            return True
        else:
            return False
                    

src/initWindow.py
import tkinter as tk
from tkinter import messagebox
from ballMoveSimulation import BallMoveSimulation


class InitWindow:
    def __init__(self, help='', version=''):
        self.__init_window = tk.Tk()
        self.__init_window.title("Начальное окно")
        self.__init_window.geometry("520x380")

        self.__version_text = version
        self.__help_text = help

        self.variables = {
            "WINDOW_WIDTH": {"value": 900, "label": "Ширина экрана"},
            "WINDOW_HEIGHT": {"value": 600, "label": "Высота экрана"},
            "BALL_RADIUS": {"value": 20, "label": "Радиус мяча"},
            "BALL_COLOR": {"value": (25, 25, 105), "label": "Цвет мяча"},
            "ball_vx": {"value": 0.0, "label": "Горизонтальная скорость мяча"},
            "ball_vy": {"value": 1.0, "label": "Вертикальная скорость мяча"},
            "ball_ax": {"value": 0.0, "label": "Горизонтальное ускорение мяча"},
            "ball_ay": {"value": 9.85, "label": "Вертикальное ускорение мяча"},
            "k": {"value": 0.8, "label": "Коэффициент упругости"},
            "clock_tick": {"value": 120, "label": "Количество кадров в секунду"},
            "escape_height": {"value": 10.0, "label": "Высота поднятия после которой программа завершается"},
        }

        self.__create_widgets()

    def __create_widgets(self):
        row = 0
        for var_name, var_info in self.variables.items():
            label_text = var_info["label"]
            default_value = var_info["value"]

            label = tk.Label(self.__init_window, text=label_text)
            label.grid(row=row, column=0, padx=4, pady=4)

            entry = tk.Entry(self.__init_window)
            entry.insert(0, str(default_value))
            entry.grid(row=row, column=1, columnspan=2,  padx=4, pady=4)
            self.variables[var_name]["entry"] = entry

            row += 1

        button_start = tk.Button(self.__init_window, text="Старт", command=self.__start_simulation)
        button_start.grid(row=row, column=0, ipadx=6, ipady=6, padx=4, pady=4)

        button_help = tk.Button(self.__init_window, text="Справка", command=self.__show_help)
        button_help.grid(row=row, column=1, ipadx=6, ipady=6, padx=4, pady=4)

        label_version = tk.Label(self.__init_window, text=self.__version_text)
        label_version.grid(row=row, column=2, ipadx=6, ipady=6, padx=4, pady=4)

    def __start_simulation(self):
        WINDOW_WIDTH = int(self.variables["WINDOW_WIDTH"]["entry"].get())
        WINDOW_HEIGHT = int(self.variables["WINDOW_HEIGHT"]["entry"].get())
        BALL_RADIUS = float(self.variables["BALL_RADIUS"]["entry"].get())
        BALL_COLOR = eval(self.variables["BALL_COLOR"]["entry"].get())
        ball_ax = float(self.variables["ball_ax"]["entry"].get())
        ball_ay = float(self.variables["ball_ay"]["entry"].get())
        ball_vx = float(self.variables["ball_vx"]["entry"].get())
        ball_vy = float(self.variables["ball_vy"]["entry"].get())
        k = float(self.variables["k"]["entry"].get())
        clock_tick = int(self.variables["clock_tick"]["entry"].get())
        escape_height = float(self.variables["escape_height"]["entry"].get())

        self.__init_window.destroy()

        simulation = BallMoveSimulation(
                WINDOW_WIDTH=WINDOW_WIDTH,
                WINDOW_HEIGHT=WINDOW_HEIGHT,
                ball_ax=ball_ax,
                ball_ay=ball_ay,
                ball_vx=ball_vx,
                ball_vy=ball_vy,
                BALL_RADIUS=BALL_RADIUS,
                BALL_COLOR=BALL_COLOR,
                k=k,
                clock_tick=clock_tick,
                escape_height=escape_height
        )
        simulation.start_simulation()

    def __show_help(self):
        messagebox.showinfo("Справка", self.__help_text)

    def run(self):
        self.__init_window.mainloop()